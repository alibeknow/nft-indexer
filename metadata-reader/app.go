package metadata_reader

import (
	"context"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/zeebo/errs"
	"go.uber.org/zap"
	"golang.org/x/sync/errgroup"

	"metadata-reader/database"
	"metadata-reader/private/httpreader"
	"metadata-reader/queue"
	"metadata-reader/server"
)

var (
	requestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "request_per_second_gauge",
			Help: "Number of requests per second",
		},
	)
	messagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "message_per_second_gauge",
			Help: "Number of messages per second",
		},
	)
	instantRequestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_request_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)

	instantMessagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_messages_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)
)

func init() {
	prometheus.MustRegister(requestPerSecond)
	prometheus.MustRegister(messagePerSecond)
	prometheus.MustRegister(instantRequestPerSecond)
	prometheus.MustRegister(instantMessagePerSecond)
}

type App struct {
	log      *zap.Logger
	db       *database.DB
	srv      *server.Server
	listener *queue.Listener
}

func NewApp(log *zap.Logger, db *database.DB, srv *server.Server, listener *queue.Listener) (*App, error) {
	app := &App{
		log:      log,
		db:       db,
		srv:      srv,
		listener: listener,
	}

	return app, nil
}

func (mr *App) Run(ctx context.Context) error {
	group, ctx := errgroup.WithContext(ctx)

	group.Go(func() error {
		now := time.Now()
		ticker := time.NewTicker(time.Second)
		defer ticker.Stop()

		var (
			reqLast int64
			msgLast int64
		)

		for {
			select {
			case <-ctx.Done():
				return nil
			case <-ticker.C:
				duration := time.Since(now).Seconds()

				reqCount := httpreader.TotalRequestCount()
				requestPerSecond.Set(float64(reqCount) / duration)
				reqDelta := reqCount - reqLast
				reqLast = reqCount
				instantRequestPerSecond.Set(float64(reqDelta))

				msgCount := queue.TotalMessageCount()
				messagePerSecond.Set(float64(msgCount) / duration)
				msgDelta := msgCount - msgLast
				msgLast = msgCount
				instantMessagePerSecond.Set(float64(msgDelta))
			}
		}
	})
	group.Go(func() error {
		return mr.srv.Run(ctx)
	})
	group.Go(func() error {
		return mr.listener.Run(ctx)
	})

	return group.Wait()
}

func (mr *App) Close() error {
	serverErr := mr.srv.Close()
	listenerErr := mr.listener.Close()

	return errs.Combine(serverErr, listenerErr)
}
