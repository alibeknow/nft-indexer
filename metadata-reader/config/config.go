package config

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"

	"github.com/spf13/viper"

	"metadata-reader/database"
	"metadata-reader/metadata"
	"metadata-reader/private/logutils"
	"metadata-reader/private/opensearch"
	"metadata-reader/queue"
	"metadata-reader/server"
)

type Config struct {
	Logger   logutils.Config
	Server   server.Config
	Database database.Config
	Queue    queue.Config
	Metadata metadata.Config
	Search   opensearch.Config
}

func LoadConfig(configPaths ...string) (Config, error) {
	v := viper.New()
	v.SetConfigName("config")
	v.SetConfigType("yaml")
	v.SetEnvPrefix("MR")
	v.SetEnvKeyReplacer(strings.NewReplacer(`.`, `_`))
	v.AutomaticEnv()

	v2 := viper.New()
	v2.SetConfigName("config")
	v2.SetConfigType("yaml")
	var buff bytes.Buffer
	m := map[string]interface{}{}
	err := json.NewEncoder(&buff).Encode(Config{})
	if err != nil {
		return Config{}, fmt.Errorf("failed to encode config: %v", err)
	}
	err = json.NewDecoder(&buff).Decode(&m)
	if err != nil {
		return Config{}, fmt.Errorf("failed to decode config: %v", err)
	}
	for key, val := range m {
		v2.Set(key, val)
	}

	for _, path := range configPaths {
		v.AddConfigPath(path)
		v2.AddConfigPath(path)
	}

	if err = v2.SafeWriteConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileAlreadyExistsError); ok {
			log.Println("config.yaml file already exists: skip creating")
		} else {
			return Config{}, fmt.Errorf("failed to write the configuration file: %v", err)
		}
	}

	if err = v.ReadInConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileNotFoundError); ok {
			log.Println("config.yaml ignored")
		} else {
			return Config{}, fmt.Errorf("failed to read the configuration file: %v", err)
		}
	}

	for _, k := range v.AllKeys() {
		value := v.GetString(k)
		v.Set(k, os.ExpandEnv(value))
	}

	var config Config
	err = v.Unmarshal(&config)
	return config, err
}
