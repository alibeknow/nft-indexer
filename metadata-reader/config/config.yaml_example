server:
  address: ":xxx"

database:
  name: "xxx"
  protocol: "mongodb"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "mongodb" for docker
  port: xxx #number
  sslEnable: true/false
  directConnection: true/false

queue:
  protocol: "amqp"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "rabbitmq" for docker
  port: xxx #number
  name: "xxx"

reader:
  workers: xxx #number
  proxyUri: "xxx"
  ipfsGateway: "https://ipfs.io/ipfs/"
  pinataDefaultGateway: "https://gateway.pinata.cloud/ipfs/"

httpreader:
  retries: xxx #number
