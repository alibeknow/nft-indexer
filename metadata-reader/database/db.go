package database

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"fmt"
	"os"
	"time"

	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
	"go.mongodb.org/mongo-driver/mongo/readpref"
	"go.mongodb.org/mongo-driver/mongo/writeconcern"
	"go.uber.org/zap"
)

type Config struct {
	Name             string
	Protocol         string
	Username         string
	Password         string
	Host             string
	Port             int
	SSLEnable        bool
	DirectConnection bool
}

func (config *Config) URL() string {
	url := fmt.Sprintf(
		"%s://%s:%s@%s:%d",
		config.Protocol,
		config.Username,
		config.Password,
		config.Host,
		config.Port,
	)

	if config.DirectConnection {
		url = fmt.Sprintf(
			"%s/?connect=direct",
			url,
		)
	}
	return url
}

type DB struct {
	log    *zap.Logger
	client *mongo.Client
	name   string
}

func Open(log *zap.Logger, ctx context.Context, config Config) (*DB, error) {
	opts := options.Client().ApplyURI(config.URL())

	if config.SSLEnable {
		path, _ := os.Getwd()
		pemPath := fmt.Sprintf("%s/../rds-combined-ca-bundle.pem", path)
		rootPEM, err := os.ReadFile(pemPath)
		if err != nil {
			return nil, err
		}

		roots := x509.NewCertPool()
		if ok := roots.AppendCertsFromPEM(rootPEM); !ok {
			return nil, fmt.Errorf("get certs from %s fail!\n", pemPath)
		}
		tlsConfig := &tls.Config{
			RootCAs:            roots,
			InsecureSkipVerify: true,
		}

		opts.SetReadPreference(readpref.Secondary())
		opts.SetWriteConcern(writeconcern.New(writeconcern.WMajority(), writeconcern.J(true), writeconcern.WTimeout(1000)))
		opts.SetTLSConfig(tlsConfig)
	}

	client, err := mongo.NewClient(opts)
	if err != nil {
		return nil, err
	}

	err = client.Connect(ctx)
	if err != nil {
		return nil, err
	}

	pingCtx, pingCancel := context.WithTimeout(ctx, 5*time.Second)
	defer pingCancel()
	err = client.Ping(pingCtx, nil)
	if err != nil {
		return nil, err
	}
	log.Info("Connected to DB", zap.String("url", config.URL()))

	return &DB{
		log:    log,
		client: client,
		name:   config.Name,
	}, nil
}

func (db *DB) Close() error {
	return db.client.Disconnect(context.Background())
}

func (db *DB) Metadata() *MetadataDB {
	coll := db.client.Database(db.name).Collection("metadata")
	return &MetadataDB{
		collection: coll,
	}
}
