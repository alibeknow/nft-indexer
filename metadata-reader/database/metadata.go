package database

import (
	"context"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"

	"metadata-reader/metadata"
)

type MetadataDB struct {
	collection *mongo.Collection
}

func (mDB MetadataDB) InsertOrUpdate(ctx context.Context, m metadata.Metadata) error {
	currentTime := time.Now()

	filter := bson.M{"_id": m.ID}
	update := bson.M{
		"$set": bson.M{
			"metadata":   m.Metadata,
			"type":       m.Type,
			"updated_at": currentTime,
		},
		"$setOnInsert": bson.M{
			"created_at": currentTime,
		},
	}

	opts := options.Update().SetUpsert(true)
	_, err := mDB.collection.UpdateOne(ctx, filter, update, opts)
	return err
}
