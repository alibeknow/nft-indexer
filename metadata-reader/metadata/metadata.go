package metadata

import (
	"context"
	"time"
)

type Metadata struct {
	ID        string     `bson:"_id"`
	Metadata  string     `bson:"metadata"`
	Type      string     `bson:"type"`
	CreatedAt *time.Time `bson:"created_at,omitempty"`
	UpdatedAt *time.Time `bson:"updated_at,omitempty"`
}

type DB interface {
	InsertOrUpdate(ctx context.Context, m Metadata) error
}
