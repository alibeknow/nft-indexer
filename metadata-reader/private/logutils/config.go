package logutils

type Config struct {
	Level string
}
