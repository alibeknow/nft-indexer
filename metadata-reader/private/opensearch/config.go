package opensearch

type Config struct {
	Uri      string
	Shards   int
	Replicas int
}
