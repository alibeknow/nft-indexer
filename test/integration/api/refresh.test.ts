import { v4 as uuid } from 'uuid';
process.env.METADATA_REPO_PROVIDER = 'mongo';
process.env.RABBIT_METADATA_READ_QUEUE = `test-queue-${uuid()}`;
process.env.COMMUNICATION_TRANSPORT = 'rmq';

import { IApiConfig, getApiConfig } from '@api/app.config';
import { RefreshModule } from '@api/refresh';
import { HttpServer, INestApplication } from '@nestjs/common';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { Test } from '@nestjs/testing';
import { Blockchain } from '@shared/blockchain';
import { ContractsRepo } from '@shared/contracts';
import { DB, DBClass, IDatabaseConfig, getDatabaseConfig } from '@shared/db';
import { Metadata, MetadataRepoDB } from '@shared/metadata';
import { IRabbitConfig } from '@shared/microservices';
import { TokenStandard, TokensRepo } from '@shared/tokens';
import * as amqplib from 'amqplib';
import { LoggerModule } from 'nestjs-pino';
import request from 'supertest';
import { getNestApp } from '../../helpers/app';
import { fixtureMetdata } from '../fixtures/fixtures.metadata';
import { fixtureApiTokens } from '../fixtures/token.api.fixture';

async function seedMetadataRepo(db: DBClass): Promise<(Metadata | null)[]> {
  const metadataRepo = new MetadataRepoDB(db);
  const contractRepo = new ContractsRepo(db);
  const tokenRepo = new TokensRepo(db);
  await contractRepo.insertIfNone({
    blockchain: Blockchain.ETH,
    address: '0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f',
    block: 15177280,
    name: 'Lost Illusion',
    type: TokenStandard.ERC721,
    createdAt: new Date(),
  });
  await Promise.all(
    fixtureApiTokens.map((token) => {
      const [ , contractAddress, tokenId ] = token._id.split(':');
      tokenRepo.save({
        block: token.block,
        blockchain: Blockchain.ETH,
        from: '0x95d35C8a511F5877Af28B515ab4f0A03B730ae9C',
        contractAddress,
        tokenId,
        count: token.count,
        createdAt: token.createdAt,
        tokenUri: token.tokenUri,
      });
    }),
  );

  return Promise.all(
    fixtureMetdata.map(({ _id, metadata, type }) =>
      metadataRepo.save(_id, metadata, type),
    ),
  );
}

jest.setTimeout(60000);

describe('[Metadata endpoints]', () => {
  let app: INestApplication;
  let httpServer: HttpServer;
  let db: DBClass;
  let amqpClient: amqplib.Channel;
  let amqpConn: amqplib.Connection;
  let metadataQueue: string;

  beforeAll(async () => {
    const dbConfig: IDatabaseConfig = getDatabaseConfig();
    db = new DBClass({ ...dbConfig, db: `api-refresh-test-${dbConfig.db}` });
    await db.open();

    await db.metadata().deleteMany({});
    await db.tokens().deleteMany({});
    await db.contracts().deleteMany({});
    await seedMetadataRepo(db);
    const moduleRef = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({
          load: [ getApiConfig ],
          isGlobal: true,
        }),
        RefreshModule,
        LoggerModule.forRoot(),
      ],
    })
      .overrideProvider(DB)
      .useValue(db)
      .compile();

    const configService = moduleRef.get<ConfigService<IApiConfig>>(
      ConfigService<IApiConfig>,
    );
    app = await getNestApp(moduleRef);
    httpServer = app.getHttpServer();

    const configRabbit = configService.get<IRabbitConfig>(
      'rabbit',
    ) as IRabbitConfig;

    const { protocol, username, password, host, port } = configRabbit;

    metadataQueue = configRabbit.metadataQueue;
    if (username && password) {
      amqpConn = await amqplib.connect(
        `${protocol}://${encodeURIComponent(username)}:${encodeURIComponent(
          password,
        )}@${host}:${port}`,
      );
    } else {
      amqpConn = await amqplib.connect(`${protocol}://${host}:${port}`);
    }
    amqpClient = await amqpConn.createChannel();
    await amqpClient.assertQueue(metadataQueue, { durable: false });
  });

  describe('[POST /api/v0/refresh/contract]', () => {
    it('should return correct result based on request params and body', async () => {
      const actualResult1 = await request(httpServer)
        .post('/api/v0/refresh/contract')
        .send({
          chainName: 'eth',
          contractAddress: '0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f',
        })
        .set('Accept', 'application/json');

      expect(actualResult1.statusCode).toBe(200);
      expect(actualResult1.body).toMatchSnapshot();
    });

    it('should return correct result based on request params and body with non-checksum formatted contract address', async () => {
      const actualResult1 = await request(httpServer)
        .post('/api/v0/refresh/contract')
        .send({
          chainName: 'eth',
          contractAddress: '0x9fb2eeb75754815c5cc9092cd53549cea5dc404f',
        })
        .set('Accept', 'application/json');

      expect(actualResult1.statusCode).toBe(200);
      expect(actualResult1.body).toMatchSnapshot();
    });
  });

  describe('[POST /api/v0/refresh/token]', () => {
    it('should return correct results based on params and body', async () => {
      const actualResult1 = await request(httpServer)
        .post('/api/v0/refresh/token')
        .send({
          chainName: 'eth',
          contractAddress: '0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f',
          tokenId: '0x04',
        })
        .set('Accept', 'application/json');

      expect(actualResult1.statusCode).toBe(200);
      expect(actualResult1.body).toMatchSnapshot();
    });

    it('should return correct results based on params and body with non-checksum formatted contract address', async () => {
      const actualResult1 = await request(httpServer)
        .post('/api/v0/refresh/token')
        .send({
          chainName: 'eth',
          contractAddress: '0X9FB2EEB75754815C5CC9092CD53549CEA5DC404F',
          tokenId: '0x04',
        })
        .set('Accept', 'application/json');

      expect(actualResult1.statusCode).toBe(200);
      expect(actualResult1.body).toMatchSnapshot();
    });

    // eslint-disable-next-line jest/no-done-callback, linebreak-style
    it('should send message to queue token API', (done: jest.DoneCallback) => {
      let counter = 0;
      request(httpServer)
        .post('/api/v0/refresh/token')
        .send({
          chainName: 'eth',
          contractAddress: '0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f',
          tokenId: '0x04',
        })
        .set('Accept', 'application/json')
        .expect(() => {
          amqpClient.consume(metadataQueue, (msg) => {
            if (msg !== null) {
              counter++;
              amqpClient.ack(msg);
              // eslint-disable-next-line jest/no-conditional-expect
              expect(counter).toBe(1);
              done();
            }
          });
        })
        .end((err) => done(err));
    });

    //eslint-disable-next-line jest/no-done-callback, linebreak-style
    it('should send message to queue contract API', (done: jest.DoneCallback) => {
      let counter = 0;
      request(httpServer)
        .post('/api/v0/refresh/contract')
        .send({
          chainName: 'eth',
          contractAddress: '0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f',
        })
        .set('Accept', 'application/json')
        .expect(() => {
          amqpClient.consume(metadataQueue, (msg) => {
            if (msg !== null) {
              counter++;
              amqpClient.ack(msg);
              // eslint-disable-next-line jest/no-conditional-expect
              expect(counter).toBe(1);
              done();
            }
          });
        })
        .end((err) => done(err));
    });
  });

  afterAll(async () => {
    await db.metadata().deleteMany({});
    await db.tokens().deleteMany({});
    await db.contracts().deleteMany({});
    await db.close(true);
    await app.close();
    await amqpConn.close();
  });
});
