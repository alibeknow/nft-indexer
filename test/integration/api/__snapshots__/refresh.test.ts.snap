// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Metadata endpoints] [POST /api/v0/refresh/contract] should return correct result based on request params and body 1`] = `
{
  "result": "The metadata sync has been initiated for contract address 0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f",
  "statusCode": 200,
}
`;

exports[`[Metadata endpoints] [POST /api/v0/refresh/contract] should return correct result based on request params and body with non-checksum formatted contract address 1`] = `
{
  "result": "The metadata sync has been initiated for contract address 0x9fb2eeb75754815c5cc9092cd53549cea5dc404f",
  "statusCode": 200,
}
`;

exports[`[Metadata endpoints] [POST /api/v0/refresh/token] should return correct results based on params and body 1`] = `
{
  "result": "The metadata sync has been initiated",
  "statusCode": 200,
}
`;

exports[`[Metadata endpoints] [POST /api/v0/refresh/token] should return correct results based on params and body with non-checksum formatted contract address 1`] = `
{
  "result": "The metadata sync has been initiated",
  "statusCode": 200,
}
`;
